In [56]:
import pandas as pd
import numpy as np

#### 0. cs데이터에서 마지막 행을 추출하시오. (loc과 iloc 모두 사용)

In [24]:
cs = pd.read_csv('dataCustomers.tab',sep='\t',encoding='cp949')
cs.head()

custid gender  age marriage     residence    job
0   10070      여   28       미혼    Yongsan-gu    제조업
1   10139      여   28       미혼   Gangdong-gu  정보서비스
2   10208      여   28       미혼   Gwangjin-gu    제조업
3   10275      여   28       미혼  Eunpyeong-gu   개인사업
4   10350      남   28       미혼    Gangnam-gu    제조업

In [224]:
## code here

cs.iloc[-1]
cs.tail()
cs.loc[499]

custid            49470
gender                남
age                  69
marriage             기혼
residence    Yongsan-gu
job                 제조업
Name: 499, dtype: object

#### 1. cs데이터에서 거주지의 비율을 구하고 cs데이터에 비율열을 추가해보세요.

In [208]:
## code here
a=pd.DataFrame(cs.residence.value_counts()/len(cs.residence))
a.columns=['rate']
a['residence']=a.index
pd.merge(cs,a, on='residence',how='left')

custid gender  age marriage      residence    job   rate
0     10070      여   28       미혼     Yongsan-gu    제조업  0.038
1     10139      여   28       미혼    Gangdong-gu  정보서비스  0.090
2     10208      여   28       미혼    Gwangjin-gu    제조업  0.058
3     10275      여   28       미혼   Eunpyeong-gu   개인사업  0.032
4     10350      남   28       미혼     Gangnam-gu    제조업  0.180
5     10425      남   28       미혼      Songpa-gu   교육기관  0.066
6     10499      여   28       미혼   Seodaemun-gu   개인사업  0.094
7     10567      여   28       미혼     Dongjak-gu  정보서비스  0.032
8     10632      여   28       미혼   Eunpyeong-gu   개인사업  0.032
9     10707      여   28       기혼     Dongjak-gu   개인사업  0.032
10    10783      여   28       미혼     Dongjak-gu   개인사업  0.032
11    10853      여   28       미혼    Jungnang-gu    제조업  0.018
12    10919      여   28       기혼  Dongdaemun-gu   개인사업  0.010
13    10983      여   28       미혼     Gangnam-gu   의료기관  0.180
14    11046      여   28       미혼   Seongdong-gu   금융기관  0.042
15    11113      여   28       미혼      Seocho-gu   의료기관  0.108
16    11182      남   28       미혼        Mapo-gu   개인사업  0.094
17    11264      여   28       미혼   Seodaemun-gu   교육기관  0.094
18    11343      여   28       기혼      Gwanak-gu   개인사업  0.018
19    11421      여   27       미혼     Gangseo-gu   의료기관  0.008
20    11489      남   29       미혼     Gangbuk-gu   개인사업  0.008
21    11563      여   28       기혼        Guro-gu    제조업  0.004
22    11631      여   28       기혼   Yangcheon-gu  운송서비스  0.028
23    11694      여   28       미혼  Dongdaemun-gu  정보서비스  0.010
24    11767      여   30       기혼      Songpa-gu    제조업  0.066
25    11843      남   28       미혼   Seodaemun-gu    그룹사  0.094
26    11905      여   29       기혼      Songpa-gu   개인사업  0.066
27    11969      남   29       미혼      Seocho-gu   개인사업  0.108
28    12035      남   29       미혼   Seodaemun-gu   의료기관  0.094
29    12105      여   29       미혼    Gangdong-gu    제조업  0.090
..      ...    ...  ...      ...            ...    ...    ...
470   46230      여   57       기혼    Jungnang-gu   개인사업  0.018
471   46327      여   58       기혼      Seocho-gu   금융기관  0.108
472   46441      여   58       기혼     Gangnam-gu   교육기관  0.180
473   46554      여   58       기혼   Seodaemun-gu  정보서비스  0.094
474   46661      남   58       기혼      Seocho-gu    제조업  0.108
475   46758      남   59       기혼    Gangdong-gu   개인사업  0.090
476   46869      여   59       기혼      Seocho-gu    건설업  0.108
477   46969      남   59       기혼    Gwangjin-gu   개인사업  0.058
478   47084      남   59       기혼     Gangnam-gu    그룹사  0.180
479   47203      남   59       기혼    Gangdong-gu   개인사업  0.090
480   47319      여   60       기혼        Jung-gu    제조업  0.016
481   47410      남   60       기혼      Seocho-gu   언론기관  0.108
482   47523      여   60       기혼   Yangcheon-gu   교육기관  0.028
483   47614      여   60       기혼   Seodaemun-gu   개인사업  0.094
484   47739      남   61       기혼     Gangnam-gu   교육기관  0.180
485   47842      여   61       기혼      Songpa-gu   개인사업  0.066
486   47956      남   61       기혼     Gangnam-gu   개인사업  0.180
487   48044      여   62       기혼       Nowon-gu    전문직  0.012
488   48169      여   62       기혼      Seocho-gu   공공기관  0.108
489   48278      여   63       기혼   Seodaemun-gu   개인사업  0.094
490   48401      여   63       기혼     Gangnam-gu   교육기관  0.180
491   48556      여   64       기혼   Seodaemun-gu   개인사업  0.094
492   48686      여   63       기혼      Songpa-gu   도소매업  0.066
493   48796      남   65       기혼      Songpa-gu    제조업  0.066
494   48890      남   65       기혼     Yongsan-gu   금융기관  0.038
495   48997      여   66       기혼     Gangnam-gu   개인사업  0.180
496   49112      여   65       기혼   Seodaemun-gu   개인사업  0.094
497   49232      여   67       기혼   Seongdong-gu   개인사업  0.042
498   49351      여   68       기혼        Mapo-gu   개인사업  0.094
499   49470      남   69       기혼     Yongsan-gu    제조업  0.038

[500 rows x 7 columns]

#### 2-1. cs데이터의 column 순서를 오름차순으로 바꾸시오.

In [29]:
## code here

cs.reindex(sorted(cs.columns),axis=1)

age  custid gender    job marriage      residence
0     28   10070      여    제조업       미혼     Yongsan-gu
1     28   10139      여  정보서비스       미혼    Gangdong-gu
2     28   10208      여    제조업       미혼    Gwangjin-gu
3     28   10275      여   개인사업       미혼   Eunpyeong-gu
4     28   10350      남    제조업       미혼     Gangnam-gu
5     28   10425      남   교육기관       미혼      Songpa-gu
6     28   10499      여   개인사업       미혼   Seodaemun-gu
7     28   10567      여  정보서비스       미혼     Dongjak-gu
8     28   10632      여   개인사업       미혼   Eunpyeong-gu
9     28   10707      여   개인사업       기혼     Dongjak-gu
10    28   10783      여   개인사업       미혼     Dongjak-gu
11    28   10853      여    제조업       미혼    Jungnang-gu
12    28   10919      여   개인사업       기혼  Dongdaemun-gu
13    28   10983      여   의료기관       미혼     Gangnam-gu
14    28   11046      여   금융기관       미혼   Seongdong-gu
15    28   11113      여   의료기관       미혼      Seocho-gu
16    28   11182      남   개인사업       미혼        Mapo-gu
17    28   11264      여   교육기관       미혼   Seodaemun-gu
18    28   11343      여   개인사업       기혼      Gwanak-gu
19    27   11421      여   의료기관       미혼     Gangseo-gu
20    29   11489      남   개인사업       미혼     Gangbuk-gu
21    28   11563      여    제조업       기혼        Guro-gu
22    28   11631      여  운송서비스       기혼   Yangcheon-gu
23    28   11694      여  정보서비스       미혼  Dongdaemun-gu
24    30   11767      여    제조업       기혼      Songpa-gu
25    28   11843      남    그룹사       미혼   Seodaemun-gu
26    29   11905      여   개인사업       기혼      Songpa-gu
27    29   11969      남   개인사업       미혼      Seocho-gu
28    29   12035      남   의료기관       미혼   Seodaemun-gu
29    29   12105      여    제조업       미혼    Gangdong-gu
..   ...     ...    ...    ...      ...            ...
470   57   46230      여   개인사업       기혼    Jungnang-gu
471   58   46327      여   금융기관       기혼      Seocho-gu
472   58   46441      여   교육기관       기혼     Gangnam-gu
473   58   46554      여  정보서비스       기혼   Seodaemun-gu
474   58   46661      남    제조업       기혼      Seocho-gu
475   59   46758      남   개인사업       기혼    Gangdong-gu
476   59   46869      여    건설업       기혼      Seocho-gu
477   59   46969      남   개인사업       기혼    Gwangjin-gu
478   59   47084      남    그룹사       기혼     Gangnam-gu
479   59   47203      남   개인사업       기혼    Gangdong-gu
480   60   47319      여    제조업       기혼        Jung-gu
481   60   47410      남   언론기관       기혼      Seocho-gu
482   60   47523      여   교육기관       기혼   Yangcheon-gu
483   60   47614      여   개인사업       기혼   Seodaemun-gu
484   61   47739      남   교육기관       기혼     Gangnam-gu
485   61   47842      여   개인사업       기혼      Songpa-gu
486   61   47956      남   개인사업       기혼     Gangnam-gu
487   62   48044      여    전문직       기혼       Nowon-gu
488   62   48169      여   공공기관       기혼      Seocho-gu
489   63   48278      여   개인사업       기혼   Seodaemun-gu
490   63   48401      여   교육기관       기혼     Gangnam-gu
491   64   48556      여   개인사업       기혼   Seodaemun-gu
492   63   48686      여   도소매업       기혼      Songpa-gu
493   65   48796      남    제조업       기혼      Songpa-gu
494   65   48890      남   금융기관       기혼     Yongsan-gu
495   66   48997      여   개인사업       기혼     Gangnam-gu
496   65   49112      여   개인사업       기혼   Seodaemun-gu
497   67   49232      여   개인사업       기혼   Seongdong-gu
498   68   49351      여   개인사업       기혼        Mapo-gu
499   69   49470      남    제조업       기혼     Yongsan-gu

[500 rows x 6 columns]

#### 2-2. cs데이터의 age를 '나이'로 job을 '직업'으로 컬럼명을 바꾸시오. 

In [31]:
## code here

cs.rename(columns={'age':'나이','job':'직업'})

custid gender  나이 marriage      residence     직업
0     10070      여  28       미혼     Yongsan-gu    제조업
1     10139      여  28       미혼    Gangdong-gu  정보서비스
2     10208      여  28       미혼    Gwangjin-gu    제조업
3     10275      여  28       미혼   Eunpyeong-gu   개인사업
4     10350      남  28       미혼     Gangnam-gu    제조업
5     10425      남  28       미혼      Songpa-gu   교육기관
6     10499      여  28       미혼   Seodaemun-gu   개인사업
7     10567      여  28       미혼     Dongjak-gu  정보서비스
8     10632      여  28       미혼   Eunpyeong-gu   개인사업
9     10707      여  28       기혼     Dongjak-gu   개인사업
10    10783      여  28       미혼     Dongjak-gu   개인사업
11    10853      여  28       미혼    Jungnang-gu    제조업
12    10919      여  28       기혼  Dongdaemun-gu   개인사업
13    10983      여  28       미혼     Gangnam-gu   의료기관
14    11046      여  28       미혼   Seongdong-gu   금융기관
15    11113      여  28       미혼      Seocho-gu   의료기관
16    11182      남  28       미혼        Mapo-gu   개인사업
17    11264      여  28       미혼   Seodaemun-gu   교육기관
18    11343      여  28       기혼      Gwanak-gu   개인사업
19    11421      여  27       미혼     Gangseo-gu   의료기관
20    11489      남  29       미혼     Gangbuk-gu   개인사업
21    11563      여  28       기혼        Guro-gu    제조업
22    11631      여  28       기혼   Yangcheon-gu  운송서비스
23    11694      여  28       미혼  Dongdaemun-gu  정보서비스
24    11767      여  30       기혼      Songpa-gu    제조업
25    11843      남  28       미혼   Seodaemun-gu    그룹사
26    11905      여  29       기혼      Songpa-gu   개인사업
27    11969      남  29       미혼      Seocho-gu   개인사업
28    12035      남  29       미혼   Seodaemun-gu   의료기관
29    12105      여  29       미혼    Gangdong-gu    제조업
..      ...    ...  ..      ...            ...    ...
470   46230      여  57       기혼    Jungnang-gu   개인사업
471   46327      여  58       기혼      Seocho-gu   금융기관
472   46441      여  58       기혼     Gangnam-gu   교육기관
473   46554      여  58       기혼   Seodaemun-gu  정보서비스
474   46661      남  58       기혼      Seocho-gu    제조업
475   46758      남  59       기혼    Gangdong-gu   개인사업
476   46869      여  59       기혼      Seocho-gu    건설업
477   46969      남  59       기혼    Gwangjin-gu   개인사업
478   47084      남  59       기혼     Gangnam-gu    그룹사
479   47203      남  59       기혼    Gangdong-gu   개인사업
480   47319      여  60       기혼        Jung-gu    제조업
481   47410      남  60       기혼      Seocho-gu   언론기관
482   47523      여  60       기혼   Yangcheon-gu   교육기관
483   47614      여  60       기혼   Seodaemun-gu   개인사업
484   47739      남  61       기혼     Gangnam-gu   교육기관
485   47842      여  61       기혼      Songpa-gu   개인사업
486   47956      남  61       기혼     Gangnam-gu   개인사업
487   48044      여  62       기혼       Nowon-gu    전문직
488   48169      여  62       기혼      Seocho-gu   공공기관
489   48278      여  63       기혼   Seodaemun-gu   개인사업
490   48401      여  63       기혼     Gangnam-gu   교육기관
491   48556      여  64       기혼   Seodaemun-gu   개인사업
492   48686      여  63       기혼      Songpa-gu   도소매업
493   48796      남  65       기혼      Songpa-gu    제조업
494   48890      남  65       기혼     Yongsan-gu   금융기관
495   48997      여  66       기혼     Gangnam-gu   개인사업
496   49112      여  65       기혼   Seodaemun-gu   개인사업
497   49232      여  67       기혼   Seongdong-gu   개인사업
498   49351      여  68       기혼        Mapo-gu   개인사업
499   49470      남  69       기혼     Yongsan-gu    제조업

[500 rows x 6 columns]

#### 3. 샤넬 신촌점의 총 매출을 구하세요. (2가지로 풀이할 것)

In [32]:
ts = pd.read_csv('dataTransactions.tab',sep='\t',encoding='cp949')
ts.head()

datetime  custid store        product  brand corner  import  \
0  2000-05-01 10:43   18313   신촌점  4104840008000     샤넬    화장품       1   
1  2000-05-01 11:00   18313   신촌점  2700000000000     식품   일반식품       0   
2  2000-05-01 11:33   27222   신촌점  4545370944500   까사미아     가구       0   
3  2000-05-01 11:43   27222   신촌점  4500860043900   대아통상     기타       0   
4  2000-05-01 11:53   27222   신촌점  4538130048700  토이플러스   문화완구       0   

   amount  installment  
0  113000            3  
1   91950            3  
2  598000            3  
3   20100            1  
4   24000            1

In [34]:
## code here

ts.amount.sum()

ts['amount'].sum()

2088404832

#### 4. ts데이터에서 한 번 구매할 때 가장 큰 지출을 보인 custid를 구하시오. 

In [42]:
## code here

ts.sort_values(by='amount',ascending=False).iloc[0,1]


15968

#### 5. house데이터에서 연도별 평균 판매액을 구하시오.

In [64]:
house = pd.read_csv('house.csv')
house.head()

id             date     price  bedrooms  bathrooms  floors  \
0  7129300520  20141013T000000  221900.0         3       1.00     1.0   
1  6414100192  20141209T000000  538000.0         3       2.25     2.0   
2  5631500400  20150225T000000  180000.0         2       1.00     1.0   
3  2487200875  20141209T000000  604000.0         4       3.00     1.0   
4  1954400510  20150218T000000  510000.0         3       2.00     1.0   

   waterfront  view  condition  grade  yr_built  
0           0     0          3      7      1955  
1           0     0          3      7      1951  
2           0     0          3      6      1933  
3           0     0          5      7      1965  
4           0     0          3      8      1987

In [81]:
## code here

y=list(set(house.yr_built))
yr={}
for i in y:
    yr[i]= house.loc[house['yr_built'] == i,'price'].mean()
yr

{1900: 581536.632183908,
 1901: 557108.3448275862,
 1902: 673192.5925925926,
 1903: 480958.1956521739,
 1904: 583867.7555555556,
 1905: 753443.9324324324,
 1906: 670027.6630434783,
 1907: 676324.4769230769,
 1908: 564499.8488372093,
 1909: 696448.9893617021,
 1910: 671671.8358208955,
 1911: 632584.2465753425,
 1912: 613193.2278481013,
 1913: 586066.2711864407,
 1914: 615246.074074074,
 1915: 585036.921875,
 1916: 601041.6202531646,
 1917: 528126.7857142857,
 1918: 492346.875,
 1919: 537887.5568181818,
 1920: 477761.0306122449,
 1921: 613224.2105263158,
 1922: 569794.1473684211,
 1923: 618653.7738095238,
 1924: 570419.928057554,
 1925: 607316.6060606061,
 1926: 625443.3777777777,
 1927: 654154.2086956522,
 1928: 621920.1984126985,
 1929: 574396.8421052631,
 1930: 600659.3111111111,
 1931: 661781.1475409836,
 1932: 458488.4210526316,
 1933: 772483.3333333334,
 1934: 517152.38095238095,
 1935: 542229.7916666666,
 1936: 642335.0,
 1937: 648791.5441176471,
 1938: 548884.9038461539,
 1939: 5

#### 6. house데이터에서 1900년대에 지어졌거나 등급이 평균 이상인 집값의 합을 구하시오. (2가지로 풀이할 것)

In [82]:
## code here

house.loc[(house['yr_built']>1900)|(house['yr_built']>house['yr_built'].mean())]

id             date      price  bedrooms  bathrooms  floors  \
0      7129300520  20141013T000000   221900.0         3       1.00     1.0   
1      6414100192  20141209T000000   538000.0         3       2.25     2.0   
2      5631500400  20150225T000000   180000.0         2       1.00     1.0   
3      2487200875  20141209T000000   604000.0         4       3.00     1.0   
4      1954400510  20150218T000000   510000.0         3       2.00     1.0   
5      7237550310  20140512T000000  1230000.0         4       4.50     1.0   
6      1321400060  20140627T000000   257500.0         3       2.25     2.0   
7      2008000270  20150115T000000   291850.0         3       1.50     1.0   
8      2414600126  20150415T000000   229500.0         3       1.00     1.0   
9      3793500160  20150312T000000   323000.0         3       2.50     2.0   
10     1736800520  20150403T000000   662500.0         3       2.50     1.0   
11     9212900260  20140527T000000   468000.0         2       1.00     1.0   
12      114101516  20140528T000000   310000.0         3       1.00     1.5   
13     6054650070  20141007T000000   400000.0         3       1.75     1.0   
15     9297300055  20150124T000000   650000.0         4       3.00     2.0   
16     1875500060  20140731T000000   395000.0         3       2.00     2.0   
17     6865200140  20140529T000000   485000.0         4       1.00     1.5   
18       16000397  20141205T000000   189000.0         2       1.00     1.0   
19     7983200060  20150424T000000   230000.0         3       1.00     1.0   
20     6300500875  20140514T000000   385000.0         4       1.75     1.0   
21     2524049179  20140826T000000  2000000.0         3       2.75     1.0   
22     7137970340  20140703T000000   285000.0         5       2.50     2.0   
23     8091400200  20140516T000000   252700.0         2       1.50     1.0   
24     3814700200  20141120T000000   329000.0         3       2.25     2.0   
25     1202000200  20141103T000000   233000.0         3       2.00     1.5   
26     1794500383  20140626T000000   937000.0         3       1.75     2.0   
27     3303700376  20141201T000000   667000.0         3       1.00     1.5   
28     5101402488  20140624T000000   438000.0         3       1.75     1.0   
29     1873100390  20150302T000000   719000.0         4       2.50     2.0   
30     8562750320  20141110T000000   580500.0         3       2.50     2.0   
...           ...              ...        ...       ...        ...     ...   
21583  2025049203  20140610T000000   399950.0         2       1.00     2.0   
21584   952006823  20141202T000000   380000.0         3       2.50     2.0   
21585  3832050760  20140828T000000   270000.0         3       2.50     2.0   
21586  2767604724  20141015T000000   505000.0         2       2.50     3.0   
21587  6632300207  20150305T000000   385000.0         3       2.50     3.0   
21588  2767600688  20141113T000000   414500.0         2       1.50     2.0   
21589  7570050450  20140910T000000   347500.0         3       2.50     2.0   
21590  7430200100  20140514T000000  1220000.0         4       3.50     1.5   
21591  4140940150  20141002T000000   572000.0         4       2.75     2.0   
21592  1931300412  20150416T000000   475000.0         3       2.25     3.0   
21593  8672200110  20150317T000000  1090000.0         5       3.75     2.0   
21594  5087900040  20141017T000000   350000.0         4       2.75     2.0   
21595  1972201967  20141031T000000   520000.0         2       2.25     3.0   
21596  7502800100  20140813T000000   679950.0         5       2.75     2.0   
21597   191100405  20150421T000000  1580000.0         4       3.25     2.0   
21598  8956200760  20141013T000000   541800.0         4       2.50     2.0   
21599  7202300110  20140915T000000   810000.0         4       3.00     2.0   
21600   249000205  20141015T000000  1540000.0         5       3.75     2.0   
21601  5100403806  20150407T000000   467000.0         3       2.50     3.0   
21602   844000965  20140626T000000   

In [85]:
house.query(" yr_built > 1900 or yr_built > yr_built.mean()")

id             date      price  bedrooms  bathrooms  floors  \
0      7129300520  20141013T000000   221900.0         3       1.00     1.0   
1      6414100192  20141209T000000   538000.0         3       2.25     2.0   
2      5631500400  20150225T000000   180000.0         2       1.00     1.0   
3      2487200875  20141209T000000   604000.0         4       3.00     1.0   
4      1954400510  20150218T000000   510000.0         3       2.00     1.0   
5      7237550310  20140512T000000  1230000.0         4       4.50     1.0   
6      1321400060  20140627T000000   257500.0         3       2.25     2.0   
7      2008000270  20150115T000000   291850.0         3       1.50     1.0   
8      2414600126  20150415T000000   229500.0         3       1.00     1.0   
9      3793500160  20150312T000000   323000.0         3       2.50     2.0   
10     1736800520  20150403T000000   662500.0         3       2.50     1.0   
11     9212900260  20140527T000000   468000.0         2       1.00     1.0   
12      114101516  20140528T000000   310000.0         3       1.00     1.5   
13     6054650070  20141007T000000   400000.0         3       1.75     1.0   
15     9297300055  20150124T000000   650000.0         4       3.00     2.0   
16     1875500060  20140731T000000   395000.0         3       2.00     2.0   
17     6865200140  20140529T000000   485000.0         4       1.00     1.5   
18       16000397  20141205T000000   189000.0         2       1.00     1.0   
19     7983200060  20150424T000000   230000.0         3       1.00     1.0   
20     6300500875  20140514T000000   385000.0         4       1.75     1.0   
21     2524049179  20140826T000000  2000000.0         3       2.75     1.0   
22     7137970340  20140703T000000   285000.0         5       2.50     2.0   
23     8091400200  20140516T000000   252700.0         2       1.50     1.0   
24     3814700200  20141120T000000   329000.0         3       2.25     2.0   
25     1202000200  20141103T000000   233000.0         3       2.00     1.5   
26     1794500383  20140626T000000   937000.0         3       1.75     2.0   
27     3303700376  20141201T000000   667000.0         3       1.00     1.5   
28     5101402488  20140624T000000   438000.0         3       1.75     1.0   
29     1873100390  20150302T000000   719000.0         4       2.50     2.0   
30     8562750320  20141110T000000   580500.0         3       2.50     2.0   
...           ...              ...        ...       ...        ...     ...   
21583  2025049203  20140610T000000   399950.0         2       1.00     2.0   
21584   952006823  20141202T000000   380000.0         3       2.50     2.0   
21585  3832050760  20140828T000000   270000.0         3       2.50     2.0   
21586  2767604724  20141015T000000   505000.0         2       2.50     3.0   
21587  6632300207  20150305T000000   385000.0         3       2.50     3.0   
21588  2767600688  20141113T000000   414500.0         2       1.50     2.0   
21589  7570050450  20140910T000000   347500.0         3       2.50     2.0   
21590  7430200100  20140514T000000  1220000.0         4       3.50     1.5   
21591  4140940150  20141002T000000   572000.0         4       2.75     2.0   
21592  1931300412  20150416T000000   475000.0         3       2.25     3.0   
21593  8672200110  20150317T000000  1090000.0         5       3.75     2.0   
21594  5087900040  20141017T000000   350000.0         4       2.75     2.0   
21595  1972201967  20141031T000000   520000.0         2       2.25     3.0   
21596  7502800100  20140813T000000   679950.0         5       2.75     2.0   
21597   191100405  20150421T000000  1580000.0         4       3.25     2.0   
21598  8956200760  20141013T000000   541800.0         4       2.50     2.0   
21599  7202300110  20140915T000000   810000.0         4       3.00     2.0   
21600   249000205  20141015T000000  1540000.0         5       3.75     2.0   
21601  5100403806  20150407T000000   467000.0         3       2.50     3.0   
21602   844000965  20140626T000000   

#### 7. house 데이터에서 월별 총 판매건수를 구하시오.

In [223]:
## code here

from collections import Counter

d=list(house.date)
dd=[]
for i in d:
    dd.append(i[4:6])

month_sale = Counter(dd);month_sale

Counter({'10': 1878,
         '12': 1471,
         '02': 1250,
         '05': 2414,
         '06': 2180,
         '01': 978,
         '04': 2231,
         '03': 1875,
         '07': 2211,
         '08': 1940,
         '11': 1411,
         '09': 1774})

#### 8. titanic데이터에서 결측치가 있는 열을 확인하고, 나이는 평균값으로 나머지 열은 '글쎄여' 로 결측값을 채우시오.

In [86]:
data = pd.read_csv('titanic.csv')
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [130]:
## code here

data.isnull().sum()
data.fillna({'Age':data.Age.mean()},inplace=True)
data.fillna('글쎄여',inplace=True)

## 화이팅^^!

In [129]:
trans = pd.read_csv('02거래정보.csv', encoding = 'cp949')
trans = trans.query('pd_c != "unknown"').loc[:,['clnt_id', 'pd_c']]

출처 : 롯데멤버스, L.pay|L.POINT, 제6회 L.POINT Big Data Competition

혹시라도 이 데이터를 불러오는 것이 답답하고 어렵다면 노트북을 바꿔야 한다는 뜻 입니다. 

In [100]:
trans.head()

clnt_id  pd_c
4    39423  0565
5    21279  0565
6    48969  0572
7    30533  0670
8    64346  0543

In [101]:
trans.shape

(582309, 2)

<br>trans 데이터는 온라인 쇼핑몰의 transaction을 보여주는 데이터다.</br>
<br>clnt_id는 고객번호를 뜻하며 고객을 구분해준다. pd_c는 상품 코드를 뜻하고 상품을 구분해준다.</br>
<br>clnt_id의 고유한 값의 개수는 11271개지만 값의 범위는 2 ~ 72424 이다.</br>
<br>pd_c의 고유한 값의 개수는 1667개지만 값의 범위는  1 ~ 1668 이다.</br>
<br>특정 패키지를 사용할 때 이 값들의 범위를 0 ~ (고유값의 개수 - 1)로 만들어 줘야 하는 경우가 생긴다.</br>
<br>이렇게 데이터를 다른 형식으로 만드는 과정을 인코딩이라고 한다.</br>

<br>[문제] trans 데이터 프레임의 clnt_id 와 pd_c를 인코딩 해서 새로운 데이터 프레임을 생성하시오</br>
<br>기존 데이터프레임과 새로 만들어진 데이터프레임의 shape는 같아야 한다.</br>

```
clnt_id = 39423 -> clnt_id = 0, pd_c = 0565 -> pd_c = 0
clnt_id = 21279 -> clnt_id = 1, pd_c = 0565 -> pd_c = 0
clnt_id = 48969 -> clnt_id = 2, pd_c = 0572 -> pd_c = 1
clnt_id = 30533 -> clnt_id = 3, pd_c = 0670 -> pd_c = 2
clnt_id = 64346 -> clnt_id = 4, pd_c = 0543 -> pd_c = 3
```
```
                                        
                                                            변경 전
```
| |clnt_id|pd_c|
|-|-|-|
|0|39423|0565|
|1|21279|0565|
|2|48969|0572|
|3|30533|0670|
|4|64346|0543|

```
                                                        clnt_id: 2 ~ 72424
                                                        pd_c: 1 ~ 1668
                                                        
                                                            변경 후
```

| |clnt_id|pd_c|
|-|-|-|
|0|0|0|
|1|1|0|
|2|2|1|
|3|3|2|
|4|4|3|

```
                                                        clnt_id: 0 ~ 11270
                                                        pd_c: 0 ~ 1666
```

In [187]:
## code here

cln=pd.DataFrame(list(set(trans['clnt_id'])))
cln.columns = ['clnt_id']
cln['c']=range(0,len(cln))
cln.head()

pdc=pd.DataFrame(list(set(trans['pd_c'])))
pdc.columns = ['pd_c']
pdc['d']=range(0,len(pdc))
pdc.head()

trans_1=pd.merge(trans,cln, on='clnt_id',how='left')
trans_2=pd.merge(trans_1,pdc, on='pd_c',how='left');trans_2
trans_2.drop(['clnt_id','pd_c'],axis=1,inplace=True)
trans_2.rename(columns={'c':'clnt_id','d':'pd_c'},inplace=True);trans_2
trans_2.shape

# 코드가 너무 더러운 답이랍니다 ㅠㅠ

(582309, 2)

In [173]:
cln=pd.DataFrame(list(set(trans['clnt_id'])))
cln.columns = ['clnt_id']
cln['c']=range(0,len(cln))
cln.head()

clnt_id  c
0    32768  0
1        2  1
2    65539  2
3    32776  3
4        9  4

In [174]:
pdc=pd.DataFrame(list(set(trans['pd_c'])))
pdc.columns = ['pd_c']
pdc['d']=range(0,len(pdc))
pdc.head()

pd_c  d
0  1192  0
1  0583  1
2  0045  2
3  1601  3
4  1343  4

In [188]:
trans_1=pd.merge(trans,cln, on='clnt_id',how='left')
trans_2=pd.merge(trans_1,pdc, on='pd_c',how='left');trans_2
trans_2.drop(['clnt_id','pd_c'],axis=1,inplace=True)
trans_2.rename(columns={'c':'clnt_id','d':'pd_c'},inplace=True);trans_2

clnt_id  pd_c
0          3152  1078
1          7642  1078
2          6078    13
3         10562   788
4         10897   604
5          6969   602
6          5870   283
7          6969   283
8          6778   979
9          6969   979
10         8159   283
11         4146   602
12         8604   911
13        10130   911
14        11118   911
15         8963   997
16         2571   997
17         8604   997
18         8604   277
19         7697  1078
20         4578  1078
21         8963  1176
22         8604   888
23         8963  1629
24        11101  1629
25          217  1463
26         6078   381
27         4778  1012
28         8604  1012
29         8963  1255
...         ...   ...
582279      702   289
582280      702  1228
582281      702   655
582282     4104   881
582283     9274  1601
582284     2580   881
582285     3521  1461
582286     4104   750
582287     6057   532
582288    10793   850
582289     9274  1156
582290     1422   101
582291    11231  1044
582292    10793  1044
582293      233   794
582294     6057  1258
582295     7338  1620
582296     1138  1044
582297     8871  1529
582298     3521    52
582299     1138  1518
582300     9914   950
582301     7659  1044
582302     7659  1258
582303     1207  1044
582304     1207  1044
582305     1207  1044
582306     1207  1044
582307     1207  1044
582308     1207  1092

[582309 rows x 2 columns]